task-maker-async/src/worker.rs

```diff
@@ -0,0 +1,566 @@
+        }
+    }
+    if let Some(sys_time_limit) = limits.sys_time {
+        if resources.sys_time > sys_time_limit.as_secs_f64() {
+            return (ExecutionStatus::SysTimeLimitExceeded, resources);
+        }
+    }
+    if let Some(wall_time_limit) = limits.wall_time {
+        if resources.wall_time > wall_time_limit.as_secs_f64() {
+            return (ExecutionStatus::WallTimeLimitExceeded, resources);
+        }
+    }
+    if let Some(memory_limit) = limits.memory {
+        if resources.memory > memory_limit {
+            return (ExecutionStatus::MemoryLimitExceeded, resources);
+        }
+    }
+    match sandbox_status.status {
+        ExitStatus::Signal(signal) => (
+            ExecutionStatus::Signal(
+                signal as u32,
+                sandbox_status
+                    .status
+                    .signal_name()
+                    .unwrap_or_else(|| "unknown signal".into()),
+            ),
+            resources,
+        ),
+        ExitStatus::ExitCode(0) => (ExecutionStatus::Success, resources),
+        ExitStatus::Killed => (ExecutionStatus::WallTimeLimitExceeded, resources),
+        ExitStatus::ExitCode(code) => (ExecutionStatus::ReturnCode(code as u32), resources),
+    }
+}
+
+async fn run_exec(tool_path: PathBuf, exec: ExecutionInfo) -> Result<ExecutionResultInfo, Error> {
+    let mut sandbox_cmd = Command::new(tool_path)
+        .arg("internal-sandbox")
+        .kill_on_drop(true)
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .spawn()?;
+    let string_config = serde_json::to_string(&exec.sandbox_config)?;
+    sandbox_cmd
+        .stdin
+        .as_mut()
+        .unwrap()
+        .write_all(string_config.as_bytes())
+        .await?;
+    let output = sandbox_cmd.wait_with_output().await?;
+
+    if !output.status.success() {
+        bail!(
+            "Sandbox process failed: {}\n{}",
+            output.status.to_string(),
+            String::from_utf8_lossy(&output.stderr)
+        );
+    }
+
+    let sandbox_result: SandboxExecutionResult = serde_json::from_slice(&output.stdout)?;
+
+    let was_killed = sandbox_result.status == ExitStatus::Killed;
+
+    let (status, resources) = status_and_resources(&sandbox_result, &exec.execution.limits);
+
+    let execution_result = ExecutionResult {
+        status,
+        was_killed,
+        was_cached: false,
+        resources,
+        // These are not set here.
+        stdout: None,
+        stderr: None,
+    };
+
+    Ok(ExecutionResultInfo {
+        dir: exec.dir,
+        execution: exec.execution,
+        result: execution_result,
+    })
+}
+
+async fn send_file_to_store(
+    store: StoreClient,
+    handle: FileSetWriteHandle,
+    file: FileSetFile,
+    local_path: PathBuf,
+) -> Result<(), Error> {
+    let mut file_to_send = File::open(local_path).await?;
+    let mut buffer = [0u8; 4096];
+
+    loop {
```

Comment:
This loop looks a lot like the one with `ProvidedFileChunkIterator`, maybe add a TODO for refactoring this (sending a file to the store).